sw/device/lib/base/memory_perftest.c

```diff
@@ -0,0 +1,216 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#define PERFTEST_BUF_LEN 1000
+
+typedef struct perf_test {
+  // A human-readable name for this particular test, e.g. "memcpy".
+  const char *label;
+
+  // Setup functions for each context buffer. These function pointers must not
+  // be NULL. The runtime of these functions will not be measured.
+  void (*setup_buf1)(uint8_t *, size_t);
+  void (*setup_buf2)(uint8_t *, size_t);
+
+  // A function that exercises the function under test, e.g. memcpy. This
+  // function pointer must not be NULL. The runtime of this function will be
+  // measured.
+  void (*func)(uint8_t *buf1, uint8_t *buf2, size_t len);
+
+  // The expected number of CPU cycles that `func` will take to run.
+  size_t expected_num_cycles;
+} perf_test_t;
+
+// Run the given `perf_test_t` and return the number of cycles it took.
+static inline uint64_t perf_test_run(const perf_test_t *test, uint8_t *buf1,
+                                     uint8_t *buf2, size_t len) {
+  CHECK(test->setup_buf1 != NULL);
+  CHECK(test->setup_buf2 != NULL);
+  CHECK(test->func != NULL);
+
+  uint64_t total_clock_cycles = 0;
+  for (size_t i = 0; i <= len; ++i) {
+    // Set up each buffer from [0..i].
+    test->setup_buf1(buf1, i);
+    test->setup_buf2(buf2, i);
+
+    uint64_t start_cycles = ibex_mcycle_read();
+    test->func(buf1, buf2, i);
+    uint64_t end_cycles = ibex_mcycle_read();
+
+    // Practically, we should never see the 64-bit cycle counter overflow. Even
+    // if it took 12 hours to run all the perf tests, the clock would have to
+    // run at 427 THz to overflow the counter (2**64 cycles / (12 * 60 * 60)
+    // seconds).
+    CHECK(start_cycles < end_cycles);
```

Comment:
Although we should never see the counters wrap, the following subtract (`end - start`) is well defined and would result in  the correct cycle count for `uint64_t` args (or for any (start,end) of the same unsigned type).

Reply:
I'd prefer to, but I also can't ever see a situation where the CHECK would fail.

Reply:
Sounds good, I'll do that in the next patchset. The CHECK is not protecting us from UB, nor is it preserving correctness if we believe that the runtime of a single test will never exceed 2**64 cycles, which seems pretty believable given my THz estimate.